### PR TITLE
[ci/docker] Pin Ray dependencies and use as constraints in CI + Docker

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

We currently install dependencies in a sequential and quasi-random order. This leads to an unstable end state, with incompatible dependencies installed next to each other. In particular, recreating this end environment is not easy - installing dependencies obtained from the environment with `pip freeze` again will lead to resolution errors. Hence it is hard to replicate the pip environment we have in our OSS docker containers for users with their own docker containers.

This PR is the first part of an effort to clean this up. In this PR, we introduce a pinned version of our requirements.txt (which includes `ray[all]`). This file is then used in all our dependency installations as a constraints file.

This ensures that our set of pinned dependencies is not violated, and users should be able to reproduce a python package environment using this file more easily.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by ollie-iterators** [comment]
The lint test failure looks like it is just moving imports around so it should be easy to do.

**Comment by ollie-iterators** [comment]
The link error should be fixed by merging in recent changes

File: python/requirements_pinned.txt [hunk 1]
```diff
@@ -0,0 +1,61 @@
+## setup.py install_requires
```
Comment (amogkam) [inline python/requirements_pinned.txt:1]: what's the plan to update this over time? Is this file generated by CI? Should updating this be added as a step in the release process?

Comment (krfricke) [reply python/requirements_pinned.txt:1]: Generally the file should be automatically tracked by dependabot (we may have to rename it for that, I'll check that shortly). But yes, I'd suggest adding this to the release process

**Comment by ollie-iterators** [comment]
More tests than link tests may be solved by merging in recent changes.

**Review by amogkam — CHANGES_REQUESTED** [review]
Thanks @krfricke! 

But from my understanding I don't think this change solves the problem that we have...here's what I think we want as a solution: 

1. `requirements.txt` should be exactly the same as `setup.py`
2. `requirements.txt`, not `requirements-pinned.txt` is what should be installed in CI. This ensures our CI is equivalent to what a user will get when they do `pip install ray[all]`  and can uncover any issues caused by new dependency breakages.
3. For nightly Docker images, we can continue to use `requirements.txt`/`ray[all]`. 
4. Only for release Docker images, we freeze the deps at the time of the branch cut, and install those pinned dependencies. This avoids us having to rebuild the release docker images with new dependencies.

**Comment by krfricke** [comment]
Yes, I had a conversation about this yesterday and it matches your proposal. I'll close this PR and open a new one that surfaces the used dependencies for now.

**Pull request closed without merging**